src/mono/mono/utils/options.c

```diff
@@ -221,3 +248,94 @@ mono_options_parse_options (const char **argv, int argc, int *out_argc, MonoErro
 	}
 	*out_argc = aindex;
 }
+
+static char *
+strncat_safe (char *destination, const char *source, char *end)
+{
+	if (destination >= end)
+		return end;
+
+	return strncat (destination, source, end - destination);
+}
+
+static char *
+strncat_option_json (char *destination, MonoOptionType type, const void *value, char *end)
+{
+	switch (type) {
+	case MONO_OPTION_BOOL:
+	case MONO_OPTION_BOOL_READONLY:
+		return strncat_safe (destination, *(gboolean*)value ? "true" : "false", end);
+
+	case MONO_OPTION_INT: {
+		char * buf = option_value_to_str (type, value);
+		char * result = strncat_safe (destination, buf, end);
+		g_free(buf);
+		return result;
+	}
+
+	case MONO_OPTION_STRING: {
+		char ch;
+		const char * src = *(char**)value;
+		if (!src || !*src)
+			return strncat_safe (destination, "\"\"", end);
+
+		destination = strncat_safe (destination, "\"", end);
+		while ((ch = *src) != 0) {
+			if (destination >= (end - 2))
+				break;
+
+			switch (ch) {
+				case '\'':
+				case '\"':
+				case '\\':
+					*destination++ = '\\';
+					*destination++ = ch;
+					break;
+				default:
+					if (ch < 32) {
+						char * buf = g_strdup_printf ("\\u%04X", ch);
+						destination = strncat_safe (destination, buf, end);
+						g_free(buf);
+					} else
+						*destination++ = *src;
+					break;
+			}
+
+			src++;
+		}
+
+		destination = strncat_safe (destination, "\"", end);
+		return destination;
+	}
+
+	default:
+		g_assert_not_reached ();
+		return NULL;
+	}
+}
+
+void
+mono_options_get_as_json (char *result_buffer, int result_buffer_size)
+{
+	char *end = result_buffer + result_buffer_size - 1,
+		*destination = result_buffer;
+	gboolean need_comma = FALSE;
+
+	*destination = 0;
+	strncpy (destination, "{\n", result_buffer_size);
```

Comment:
what do you think about using `GString` and functions like `g_string_append_printf` to create a temporary result and then copy it into `destination`?

Reply:
Sounds good, I dug around in the headers and didn't find a g_ API that looked relevant for this situation but that might be a good one. I'll give it a look if you want me to rework it to use it.